operator/pkg/gateway-api/grpcroute.go

```diff
@@ -104,15 +117,74 @@ func getBackendServiceForGRPCRoute(rawObj client.Object) []string {
 			)
 		}
 	}
+	if !routechecks.HasServiceImportCRD(context.Background(), r.Client) {
```

Comment:
why not handling the `ServiceImport` in the existing loop and move the backend service lookup for a `ServiceImport` into the helper package (new `serviceimport.go` file)? (the same applies for grpc, http & tls route - hence we could remove at least some of the boilerplate code)

e.g.

```
	for _, rule := range route.Spec.Rules {
		for _, backend := range rule.BackendRefs {
			backendServiceName := ""

			switch {
			case helpers.IsService(backend.BackendObjectReference):
				backendServiceName = string(backend.Name)

			case routechecks.HasServiceImportCRD(context.Background(), r.Client) &&
				helpers.IsServiceImport(backend.BackendObjectReference):

                                svcName, err := helpers.GetBackendServiceForServiceImport(...)
                                // errorhandling: continue with logging

                                backendServiceName = svcName
			default:
				continue
			}

			backendServices = append(backendServices,
				types.NamespacedName{
					Namespace: helpers.NamespaceDerefOr(backend.Namespace, route.Namespace),
					Name:      backendServiceName,
				}.String(),
			)
		}
	}

```

wdyt?

edit: we could even re-use parts of the extracted lookup helper function in the routecheck `CheckBackendIsExistingServiceImport`

Reply:
> why not handling the ServiceImport in the existing loop

I think I was concerned at the beginning about the cost of the `HasServiceImportCRD` check but later on I made it lazy anyway yes indeed it can be inlined in the same loop thanks :+1:.

Also about the helper function indeed seems nice suggestion also, thanks! I will try to do the same as well for populating `backendServiceImports`.

Reply:
So I did some refactor to factorize some of this logic, it is not as good as I expected as I didn't recalled because of the different http,grpc,tls types that I forgot about when I was answering you but it should be better than before. I also adopted the style you suggested about the switch case, let me know wdyt!

Reply:
Thanks! Yes, we need some further abstraction for these types to have some more progress in removing more boilerplate code - but this should't matter for this PR.

It looks like you moved the whole `switch` statement into `helpers/serviceimport.go`. As outlined in my previous comment, I would only move the serviceimport related code. Otherwise k8s service (and potentially all upcoming backend ref types) have to be handled within `serviceimport.go`.

Reply:
Alternatively we could introduce another helper `servicelookup.go` (with the current function including the switch statement) - that potentially calls out to other helper functions (e..g the one to lookup a backend service for a serviceimport in `serviceimport.go`).

Reply:
Ah sure I got confused by the fact that you were suggesting two functions and I understood one, I think I did what you were actually referring to now.